### ISSUE TITLE
fd 8.1.0 fails to run (Ubuntu 18.04 - WSL)

### ISSUE THREAD

**Issue by lc9er** [Homebrew/linuxbrew-core#20393]
# Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.

- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
No logs.
- [x] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?
❯ brew config
HOMEBREW_VERSION: 2.2.17-61-gdee3903
ORIGIN: https://github.com/Homebrew/brew
HEAD: dee3903b575286fabe7383181e740c2ca4c5fe74
Last commit: 4 hours ago
Core tap ORIGIN: https://github.com/Homebrew/linuxbrew-core
Core tap HEAD: fa0f4675049f875550b2c478479316a2c668b6d7
Core tap last commit: 3 hours ago
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_MAKE_JOBS: 8
CPU: octa-core 64-bit icelake
Homebrew Ruby: 2.6.3 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/bin/ruby
Clang: N/A
Git: 2.17.1 => /usr/bin/git
Curl: 7.58.0 => /usr/bin/curl
Kernel: Linux 4.4.0-18362-Microsoft x86_64 GNU/Linux
OS: Unknown
Host glibc: 2.27
/usr/bin/gcc: 7.5.0
glibc: N/A
gcc: N/A
xorg: N/A
❯ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  guile@2

## What you were trying to do (and why)

search for files with fd.  

## What happened (include command output)

```
$ fd txt$
```
<details>
  <summary>Command output</summary>
  <pre>

  "bash: /home/linuxbrew/.linuxbrew/bin/fd: cannot execute binary file: Exec format error"

  </pre>
</details>

## What you expected to happen

return all txt files

## Step-by-step reproduction instructions (by running `brew install` commands)

brew install fd
fd txt$

**Comment by jonchang** [comment]
What's the output of `brew info fd` and `file /home/linuxbrew/.linuxbrew/bin/fd`? I can't reproduce this issue.

**Comment by lc9er** [comment]
```
❯ brew info fd
fd: stable 8.1.0 (bottled), HEAD
Simple, fast and user-friendly alternative to find
https://github.com/sharkdp/fd
/home/linuxbrew/.linuxbrew/Cellar/fd/8.1.0 (11 files, 3.2MB) *
  Poured from bottle on 2020-05-21 at 04:55:51
From: https://github.com/Homebrew/linuxbrew-core/blob/master/Formula/fd.rb
==> Dependencies
Build: rust ✘
==> Options
--HEAD
        Install HEAD version
==> Caveats
Bash completion has been installed to:
  /home/linuxbrew/.linuxbrew/etc/bash_completion.d

zsh completions have been installed to:
  /home/linuxbrew/.linuxbrew/share/zsh/site-functions
==> Analytics
install: 529 (30 days), 1,590 (90 days), 3,304 (365 days)
install-on-request: 528 (30 days), 1,590 (90 days), 3,304 (365 days)
build-error: 0 (30 days)
```
```
❯ file /home/linuxbrew/.linuxbrew/bin/fd
/home/linuxbrew/.linuxbrew/bin/fd: symbolic link to ../Cellar/fd/8.1.0/bin/fd
```

**Comment by dawidd6** [comment]
Can't reproduce the issue either.

**Comment by studoot** [comment]
I'm getting the same issue. The other thing I've tried that looks odd is that `ldd` fails when run on the fd executable:

```
stu@bean:~/.local/bin$ ldd /home/linuxbrew/.linuxbrew/bin/../Cellar/fd/8.1.0/bin/fd
ldd: exited with unknown exit code (139)
```

**Comment by studoot** [comment]
TL;DR - `fd` (Homebrew, bottled) appears to not work for me because of something in Windows Subsystem for Linux version 1. Either building from source or using WSL version 2 yield a correctly functioning `fd` executable.

---

I'm now getting this same issue with `fd` v8.1.1 from Homebrew (I did with v8.1.0 as well)... So, time for some investigation...

1.  I built `fd-8.1.1` from source using Rust 1.43.1 (unpacked the source and did `cargo run --release`) & that runs fine on my setup (which is running Ubuntu 18.04 within Windows using Windows Subsystem for Linux (WSL)) and can be processed by `ldd` (see [my previous comment](https://github.com/Homebrew/linuxbrew-core/issues/20393#issuecomment-634708830) about `ldd` not liking the `fd` executable).

2.  I then created an Ubuntu 18.04 VM under Hyper-V (the Windows VM subsystem) & copied the `fd` executable downloaded from Homebrew into it. The executable runs fine, but `ldd` still won't load it. So - this shows a discrepancy between my Ubuntu under WSL and a (close as) bare-metal Ubuntu distro.

3.  I then tried something a bit different - I downloaded and installed the Debian distro for WSL, but I installed it using version 2 of WSL, which uses a different mechanism for providing Linux syscalls than version 1, which I use for my Ubuntu distro - [this post](https://ulsoy.org/blog/experiencing-wsl-as-a-linux-veteran-part-1/) gives a good overview.

    I got Homebrew onto the Debian distro & installed `fd` with it and `fd` worked as expected. I than converted the Debian distro to use version 1 of WSL (you can convert WSL distros between WSL versions pretty much seamlessly) and `fd` no longer worked, giving the same error (`cannot execute binary file: Exec format error`) as on my Ubuntu distro. I then converted the Debian distro back to version 2 of WSL and `fd` worked again.

4. Just for completion, I then did a `brew install --reinstall -s fd` to get Homebrew to build `fd` from source. That resulted in an ELF64 executable that executed under Ubuntu/WSL1. One difference between this executable and the one I built from source 

At this point, I'm calling it a day. I think (for my particular circumstances) the issue looks like it's in version 1 of WSL, and (I guess) its incomplete emulation of Linux syscalls, rather than within `fd` or Homebrew - it appears that `execve` in WSL1 won't recognise the bottled `fd` as an ELF64 executable. The only thing that I wonder about is whether the build process has changed - I'm sure `fd` before 8.1.0 *used* to work on my WSL1 distro...

**Comment by tueda** [comment]
I hit the same issue
```bash
bash: /home/linuxbrew/.linuxbrew/bin/fd: cannot execute binary file: Exec format error
```
but `brew reinstall --build-from-source fd` worked well.

<details><summary>Environment</summary>
<p>

```
$ brew info fd | head -5
fd: stable 8.1.1 (bottled), HEAD
Simple, fast and user-friendly alternative to find
https://github.com/sharkdp/fd
/home/linuxbrew/.linuxbrew/Cellar/fd/8.1.1 (11 files, 3.2MB) *
  Poured from bottle on 2020-08-07 at 05:21:04
$ brew --version
Homebrew 2.4.9-293-g56ed27a
Homebrew/linuxbrew-core (git revision 4c7841; last commit 2020-08-05)
$ cat /etc/os-release | grep PRETTY_NAME
PRETTY_NAME="Ubuntu 18.04.4 LTS"
$ uname -srvm
Linux 4.4.0-18362-Microsoft #836-Microsoft Mon May 05 16:04:00 PST 2020 x86_64
```

</p>
</details>
